### ISSUE TITLE
Can't compile and run example on ubuntu

### ISSUE THREAD

**Issue by ArtemLubimov** [redis/hiredis#654]
Hi!. I have a problem. I'm using Ubuntu 16.04 and try to compile program with this library. I run "make -f ./Makefile" and then "make -f ./Makefile install" and after that I'll got this:
`mkdir -p /usr/local/include/hiredis /usr/local/include/hiredis/adapters /usr/local/lib
cp -pPR hiredis.h async.h read.h sds.h sslio.h /usr/local/include/hiredis
cp -pPR adapters/*.h /usr/local/include/hiredis/adapters
cp -pPR libhiredis.so /usr/local/lib/libhiredis.so.0.14
cd /usr/local/lib && ln -sf libhiredis.so.0.14 libhiredis.so
cp -pPR libhiredis.a /usr/local/lib
mkdir -p /usr/local/lib/pkgconfig
cp -pPR hiredis.pc /usr/local/lib/pkgconfig
`
When I run test.sh all test is passed. When I try to compile example.c using command - "gcc ./example.c -o example", I'll get this error:
`./example.c:5:21: fatal error: hiredis.h: No such file or directory
 #include <hiredis.h>
` 
After that I'll try to change "#include <hiredis.h>" to "#include <hiredis/hiredis.h>" and after compile I'll get:
`sudo gcc ./example.c -o ./ex
/tmp/ccpjyKVO.o: In function main:
example.c:(.text+0xe0): undefined reference to redisConnectUnixWithTimeout
example.c:(.text+0x100): undefined reference to redisConnectWithTimeout
example.c:(.text+0x142): undefined reference to redisFree
example.c:(.text+0x16e): undefined reference to redisCommand
example.c:(.text+0x198): undefined reference to freeReplyObject
example.c:(.text+0x1b8): undefined reference to redisCommand
example.c:(.text+0x1e2): undefined reference to freeReplyObject
example.c:(.text+0x20e): undefined reference to redisCommand
example.c:(.text+0x238): undefined reference to freeReplyObject
example.c:(.text+0x24e): undefined reference to redisCommand
example.c:(.text+0x278): undefined reference to freeReplyObject
example.c:(.text+0x28e): undefined reference to redisCommand
example.c:(.text+0x2b8): undefined reference to reeReplyObject
example.c:(.text+0x2ce): undefined reference to redisCommand
example.c:(.text+0x2f8): undefined reference to freeReplyObject
example.c:(.text+0x30e): undefined reference to redisCommand
example.c:(.text+0x31e): undefined reference to freeReplyObject
example.c:(.text+0x361): undefined reference to redisCommand
example.c:(.text+0x371): undefined reference to freeReplyObject
example.c:(.text+0x391): undefined reference to redisCommand
example.c:(.text+0x3f6): undefined reference to reeReplyObject
example.c:(.text+0x402): undefined reference to redisFree
collect2: error: ld returned 1 exit status
`
What I doing wrong?

**Comment by michael-grunder** [comment]
Hi, @ArtemLubimov,

It was kind of hard to follow the issue since it's kind of squished up linker errors, but I think I can help.

Try to do the following:

```bash
$ git clone https://github.com/redis/hiredis.git
$ make
$ sudo make install
$ make test
```

Once installed, save the following program to say `hiredis-test.c`

```
#include <stdio.h>
#include <hiredis/hiredis.h>

int main(void) {
    redisContext *context;
    redisReply *reply;

    context = redisConnect("127.0.0.1", 6379);
    if (!context) {
        fprintf(stderr, "Error:  Can't connect to Redis\n");
        return -1;
    }

    reply = redisCommand(context, "SET %s %s", "foo", "bar");
    if (!reply || context->err) {
        fprintf(stderr, "Error:  Can't send command to Redis\n");
        return -1;
    }

    freeReplyObject(reply);

    reply = redisCommand(context, "GET foo");
    if (!reply || context->err || reply->type != REDIS_REPLY_STRING) {
        fprintf(stderr, "Error:  Can't send command to Redis\n");
        return -1;
    }

    printf("foo = %s\n", reply->str);
    freeReplyObject(reply);

    return 0;
}
```

Then, see if you can compile and run it like so:

```bash
$ cc hiredis-example.c -o hiredis-test -lhiredis
./hiredis-test
```

**Comment by ArtemLubimov** [comment]
Thank you very much, you save my time!) The main problem in my path was in this - "-lhiredis"

**Comment by michael-grunder** [comment]
Glad you got it working!

**Comment by ashwani29** [comment]
@michael-grunder  not working for me.
compiling with `-lhiredis` is again throwing the same error
> Hi, @ArtemLubimov,
> 
> It was kind of hard to follow the issue since it's kind of squished up linker errors, but I think I can help.
> 
> Try to do the following:
> 
> ```shell
> $ git clone https://github.com/redis/hiredis.git
> $ cd hiredis
> $ make
> $ sudo make install
> $ make test
> ```
> 
> Once installed, save the following program to say `hiredis-test.c`
> 
> ```
> #include <stdio.h>
> #include <hiredis/hiredis.h>
> 
> int main(void) {
>     redisContext *context;
>     redisReply *reply;
> 
>     context = redisConnect("127.0.0.1", 6379);
>     if (!context) {
>         fprintf(stderr, "Error:  Can't connect to Redis\n");
>         return -1;
>     }
> 
>     reply = redisCommand(context, "SET %s %s", "foo", "bar");
>     if (!reply || context->err) {
>         fprintf(stderr, "Error:  Can't send command to Redis\n");
>         return -1;
>     }
> 
>     freeReplyObject(reply);
> 
>     reply = redisCommand(context, "GET foo");
>     if (!reply || context->err || reply->type != REDIS_REPLY_STRING) {
>         fprintf(stderr, "Error:  Can't send command to Redis\n");
>         return -1;
>     }
> 
>     printf("foo = %s\n", reply->str);
>     freeReplyObject(reply);
> 
>     return 0;
> }
> ```
> 
> Then, see if you can compile and run it like so:
> 
> ```shell
> $ cc hiredis-test.c -o hiredis-test -lhiredis
> ./hiredis-test
> ```

**Comment by michael-grunder** [comment]
It's hard to say.  Can you paste the actual error you're getting?

**Comment by ashwani29** [comment]
yes sure, this is the error compilation throws:
```
example.c:4:10: fatal error: hiredis.h: No such file or directory
    4 | #include <hiredis.h>
```

If you want to answer it here: https://stackoverflow.com/questions/64771884/custom-redis-client-compilation-failing

**Comment by michael-grunder** [comment]
> #include <hiredis.h>

By default hiredis will store headers in `/usr/local/include/hiredis/` so you'd need to include `<hiredis/hiredis.h>`.

I'm guessing this program will compile:

```c
#include <hiredis/hiredis.h>

int main(void) {
    redisContext *c = redisConnect("localhost", 6379);
    return 0;
}
```

Your other option would be to add `-I/usr/local/include/hiredis` to your build command.

```bash
# As an example
$ cc -I/usr/local/include/hiredis your-program.c -lhiredis
```

**Comment by ashwani29** [comment]
thanks, got it working.

**Comment by michael-grunder** [comment]
@ashwani29 It's hard to say without an example program.

**Comment by ashwani29** [comment]
This is the redis client file [redis_client.c.txt](https://github.com/redis/hiredis/files/5526394/redis_client.c.txt), redis server is running in a `runc` container. I exported the docker redis image to a file system that a runc container can use.  see here https://chromium.googlesource.com/external/github.com/docker/containerd/+/refs/tags/v0.2.2/docs/bundle.md

**Comment by michael-grunder** [comment]
The program runs just fine for me.

It did return `NULL` for the `GET foo` test but that's because it never `SET foo` to anything.

Here's a smaller program that just tries a `SET` and then a `GET`.
https://gist.github.com/michael-grunder/cf88b3217a794ca30a459743ac0974d7

You should also confirm that you can get and set keys via `redis-cli` as that also uses hiredis.

**Comment by ashwani29** [comment]
Thanks, May be the redis server running in the container is not able to save those entries in the database and btw, I used a while loop where the key was an integer and value was also an int. You sure, all that was working correctly.

**Comment by 1336258176** [comment]
@michael-grunder 
Hello,I think I'm in trouble.The following is the error I encountered, please help me to see.Thinks very much.
`/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /lib/../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/../../../libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: cannot find -lMVSDK
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/HEU_Vision_2022_MDVS.dir/build.make:341: HEU_Vision_2022_MDVS] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/HEU_Vision_2022_MDVS.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`

**Comment by 1336258176** [comment]
```
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /lib/../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/../lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/9/../../../libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: skipping incompatible /usr/lib/libMVSDK.so when searching for -lMVSDK
/usr/bin/ld: cannot find -lMVSDK
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/HEU_Vision_2022_MDVS.dir/build.make:341: HEU_Vision_2022_MDVS] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/HEU_Vision_2022_MDVS.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

**Comment by michael-grunder** [comment]
I'm not sure honestly.  I'm not familiar with `libMVSDK` but hiredis doesn't use it.

Are you trying to build a different project that builds hiredis as a part of the build?   From the build log it looks like you may have the wrong version (e.g. ARM vs x86-64) of the library in your path?

**Comment by 1336258176** [comment]
I reinstalled a different version of `MVSDK`. Then, this problem was solved. Indeed, as you said, the version was wrong. Thank you. Looking forward to communicating with you next time.
